components/omega/src/base/Halo.h

```diff
@@ -197,61 +217,46 @@ class Halo {
-   int packBuffer(const HostArray3DR4 Array);
-   int packBuffer(const HostArray3DR8 Array);
-   int packBuffer(const HostArray4DI4 Array);
-   int packBuffer(const HostArray4DI8 Array);
-   int packBuffer(const HostArray4DR4 Array);
-   int packBuffer(const HostArray4DR8 Array);
-   int packBuffer(const HostArray5DI4 Array);
-   int packBuffer(const HostArray5DI8 Array);
-   int packBuffer(const HostArray5DR4 Array);
-   int packBuffer(const HostArray5DR8 Array);
-
-   /// Buffer unpack functions overloaded to each supported Kokkos array type.
-   /// After receiving a message from a neighboring task, save the elements
-   /// of RecvBuffer for that Neighbor into the corresponding halo elements
-   /// of the input Array
-   int unpackBuffer(HostArray1DI4 &Array);
-   int unpackBuffer(HostArray1DI8 &Array);
-   int unpackBuffer(HostArray1DR4 &Array);
-   int unpackBuffer(HostArray1DR8 &Array);
-   int unpackBuffer(HostArray2DI4 &Array);
-   int unpackBuffer(HostArray2DI8 &Array);
-   int unpackBuffer(HostArray2DR4 &Array);
-   int unpackBuffer(HostArray2DR8 &Array);
-   int unpackBuffer(HostArray3DI4 &Array);
-   int unpackBuffer(HostArray3DI8 &Array);
-   int unpackBuffer(HostArray3DR4 &Array);
-   int unpackBuffer(HostArray3DR8 &Array);
-   int unpackBuffer(HostArray4DI4 &Array);
-   int unpackBuffer(HostArray4DI8 &Array);
-   int unpackBuffer(HostArray4DR4 &Array);
-   int unpackBuffer(HostArray4DR8 &Array);
-   int unpackBuffer(HostArray5DI4 &Array);
-   int unpackBuffer(HostArray5DI8 &Array);
-   int unpackBuffer(HostArray5DR4 &Array);
-   int unpackBuffer(HostArray5DR8 &Array);
+   /// Allocate the recieve buffers and call MPI_Irecv for each Neighbor.
+   /// The input bool UseDevBuffer specifies whether or not the device buffer
+   /// will be used in the unpackBuffer functionfor unpacking into the array.
+   int startReceives(bool UseDevBuffer);
+
+   /// Call MPI_Isend for each Neighbor to send the packed buffers to the
+   /// neighboring tasks. The input bool UseDevBuffer specifies whether or not
+   /// the device buffer was packed in the packBuffer function.
+   int startSends(bool UseDevBuffer);
+
+   /// Function template to find the memory space in which the input Array
+   /// is defined.
+   template <typename T> ArrayMemLoc findMemLoc(const T &Array) {
+      if (std::is_same_v<MemSpace, HostMemSpace>) {
+         return ArrayMemLoc::Both;
+      } else if (T::is_hostspace) {
+         return ArrayMemLoc::Host;
+      } else {
+         return ArrayMemLoc::Device;
+      }
+   }
+
+   /// Function template that returns a bool that is true if the Array is
+   /// on the device, or if the device and host memory spaces are the same
+   /// space. Used to determine if buffer packs and unpacks are done within
+   /// Kokkos parallelFor kernels.
+   template <typename T> bool devBufferPUP(const T &Array) {
+      bool OnDev = findMemLoc(Array) == ArrayMemLoc::Both ||
+                   findMemLoc(Array) == ArrayMemLoc::Device;
```

Comment:
Assuming that the above findMemLoc returns ArrayMemLoc::Both for a HostMemSpace view, this may cause unintentional work due to the result being true.